### PR TITLE
UCT/IB/ATOMICS: adapt IB verbs transports to new API

### PR DESCRIPTION
**Pull request by hoopoepg**
- deprecated calls are re-implemented using new API

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6735/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6742/ for details (Mellanox internal link).

File: src/uct/ib/dc/verbs/dc_verbs.c [hunk 1]
```diff
@@ -447,48 +447,131 @@ uct_dc_verbs_ep_ext_atomic(uct_dc_verbs_ep_t *ep, int opcode, void *result,
...
+                                           uint64_t remote_addr, uct_rkey_t rkey)
 {
-
     uct_dc_verbs_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_dc_verbs_iface_t);
-    uct_dc_verbs_ep_t *ep = ucs_derived_of(tl_ep, uct_dc_verbs_ep_t);
+    uct_dc_verbs_ep_t *ep       = ucs_derived_of(tl_ep, uct_dc_verbs_ep_t);
     uct_rc_iface_send_desc_t *desc;
 
-    /* TODO don't allocate descriptor - have dummy buffer */
+    if (opcode != UCT_ATOMIC_OP_ADD ) {
```
Comment (yosefe) [inline src/uct/ib/dc/verbs/dc_verbs.c]: extra space before )

File: src/uct/ib/dc/verbs/dc_verbs.c [hunk 2]
```diff
@@ -447,48 +447,131 @@ uct_dc_verbs_ep_ext_atomic(uct_dc_verbs_ep_t *ep, int opcode, void *result,
...
 
+ucs_status_t uct_dc_verbs_ep_atomic32_post(uct_ep_h tl_ep, unsigned opcode, uint32_t value,
+                                           uint64_t remote_addr, uct_rkey_t rkey)
+{
+#if HAVE_IB_EXT_ATOMICS
+    uct_dc_verbs_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_dc_verbs_iface_t);
+    uct_dc_verbs_ep_t *ep       = ucs_derived_of(tl_ep, uct_dc_verbs_ep_t);
+    uct_rc_iface_send_desc_t *desc;
+
+    if (opcode != UCT_ATOMIC_OP_ADD ) {
```
Comment (yosefe) [inline src/uct/ib/dc/verbs/dc_verbs.c]: extra space before )

File: src/uct/ib/dc/verbs/dc_verbs.c [hunk 3]
```diff
@@ -447,48 +447,131 @@ uct_dc_verbs_ep_ext_atomic(uct_dc_verbs_ep_t *ep, int opcode, void *result,
...
+                                      remote_addr, rkey, comp);
+#if HAVE_IB_EXT_ATOMICS
+    case UCT_ATOMIC_OP_SWAP:
+        return uct_dc_verbs_ep_ext_atomic(ucs_derived_of(tl_ep, uct_dc_verbs_ep_t),
+                                          IBV_EXP_WR_EXT_MASKED_ATOMIC_CMP_AND_SWP,
+                                          result, sizeof(uint64_t), 0, 0, value, remote_addr,
+                                          rkey, comp);
+#endif
+    default:
+        break;
```
Comment (yosefe) [inline src/uct/ib/dc/verbs/dc_verbs.c]: maybe just put "return UCS_ERR_UNSUPPORTED" here, without assert?

File: src/uct/ib/rc/verbs/rc_verbs_ep.c [hunk 4]
```diff
@@ -361,45 +361,131 @@ ucs_status_t uct_rc_verbs_ep_am_zcopy(uct_ep_h tl_ep, uint8_t id, const void *he
...
+                                               uint64_t remote_addr, uct_rkey_t rkey,
+                                               uct_completion_t *comp)
+{
+#if HAVE_IB_EXT_ATOMICS
+    int op;
+    uint32_t add;
+    uint32_t swap;
+
+    switch (opcode) {
+    case UCT_ATOMIC_OP_ADD:
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_ep.c]: can we unite this code with dc?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6747/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

File: src/uct/ib/dc/verbs/dc_verbs.c (hunk 3 above)
Comment (yosefe) [reply src/uct/ib/dc/verbs/dc_verbs.c]: no need to have "return" in the end of the function, can just do
```
default:
    return UCS_ERR_UNSUPPORTED;
```

File: src/uct/ib/rc/verbs/rc_verbs_common.h [hunk 5]
```diff
@@ -689,6 +689,28 @@ uct_rc_verbs_fill_ext_atomic_wr(struct ibv_exp_send_wr *wr, struct ibv_sge *sge,
...
+    case UCT_ATOMIC_OP_SWAP:
+        *op   = IBV_EXP_WR_EXT_MASKED_ATOMIC_CMP_AND_SWP;
+        *add  = 0;
+        *swap = value;
+        break;
+    default:
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return UCS_OK;
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_common.h]: no need to have "return" here, can just "return UCS_OK;" instead of "break;" in 2 places

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6753/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6763/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6765/ for details (Mellanox internal link).

**Pull request merged**